File: src/reranking/mlx_reranker.py
#!/usr/bin/env python3
"""
MLX-based reranker using Jina Reranker V3 MLX
This script runs locally on Apple Silicon (M1/M2/M3) using the MLX framework.
"""

import sys
import json
import argparse
from pathlib import Path

try:
    import mlx.core as mx
    from mlx_lm import load, generate
except ImportError as e:
    print(json.dumps({"error": f"MLX dependencies not installed: {e}"}), file=sys.stderr)
    sys.exit(1)


def load_model(model_path: str):
    """Load the MLX reranker model."""
    try:
        # Try to load the model from the specified path
        # For Jina Reranker V3 MLX, we would typically use a specialized loading function
        # This is a placeholder - actual implementation depends on the model format
        model, tokenizer = load(model_path)
        return model, tokenizer
    except Exception as e:
        raise RuntimeError(f"Failed to load model from {model_path}: {e}")


def rerank_documents(query: str, documents: list, model_path: str, top_k: int = 10):
    """
    Rerank documents based on query relevance using MLX.
    
    Args:
        query: The search query
        documents: List of document strings to rerank
        model_path: Path to the MLX model
        top_k: Number of top results to return
    
    Returns:
        List of dictionaries with 'index' and 'score' keys
    """
    try:
        model, tokenizer = load_model(model_path)
        
        # For a proper reranker implementation, we would:
        # 1. Encode the query and documents
        # 2. Compute similarity scores
        # 3. Return top-k results with scores
        
        # Placeholder implementation using simple text similarity
        # In production, this would use the actual MLX reranker model
        results = []
        for idx, doc in enumerate(documents):
            # Simple similarity score (placeholder)
            # Real implementation would use the model's scoring mechanism
            query_words = set(query.lower().split())
            doc_words = set(doc.lower().split())
            overlap = len(query_words & doc_words)
            total = len(query_words | doc_words)
            score = overlap / total if total > 0 else 0.0
            
            results.append({
                "index": idx,
                "score": float(score)
            })
        
        # Sort by score descending and take top_k
        results.sort(key=lambda x: x["score"], reverse=True)
        return results[:top_k]
        
    except Exception as e:
        raise RuntimeError(f"Reranking failed: {e}")


def main():
    """Main entry point for the MLX reranker script."""
    parser = argparse.ArgumentParser(description="MLX-based document reranker")
    parser.add_argument("--query", required=True, help="Search query")
    parser.add_argument("--model", required=True, help="Path to MLX model")
    parser.add_argument("--top-k", type=int, default=10, help="Number of top results to return")
    parser.add_argument("--documents", required=True, help="JSON array of documents to rerank")
    
    args = parser.parse_args()
    
    try:
        # Parse documents from JSON
        documents = json.loads(args.documents)
        
        if not isinstance(documents, list):
            raise ValueError("Documents must be a JSON array")
        
        # Perform reranking
        results = rerank_documents(
            query=args.query,
            documents=documents,
            model_path=args.model,
            top_k=args.top_k
        )
        
        # Output results as JSON
        print(json.dumps({"results": results}))
        
    except json.JSONDecodeError as e:
        print(json.dumps({"error": f"Invalid JSON in documents: {e}"}), file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(json.dumps({"error": str(e)}), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/reranking/mlx-reranker.ts
/**
 * MLX-based reranker implementation
 * 
 * This reranker uses Apple's MLX framework to run Jina Reranker V3 MLX locally
 * on Apple Silicon (M1/M2/M3) chips. It provides fast, local reranking without
 * API calls by spawning a Python subprocess to run the MLX model.
 * 
 * Requirements:
 * - Apple Silicon (M1/M2/M3) Mac
 * - Python 3.8+ with MLX installed: pip install mlx mlx-lm
 * - Jina Reranker V3 MLX model downloaded locally
 */

import { spawn } from 'child_process';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import type { Reranker, RerankerConfig, RerankOptions, RerankResult } from '../types.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * Configuration for MLX reranker
 */
interface MlxRerankerConfig extends RerankerConfig {
    /** Path to the MLX model directory */
    modelPath: string;
    /** Path to Python executable (default: 'python3') */
    pythonPath?: string;
}

/**
 * Response from the Python MLX reranker script
 */
interface MlxRerankerResponse {
    results?: Array<{
        index: number;
        score: number;
    }>;
    error?: string;
}

/**
 * MLX-based reranker implementation
 * 
 * This class implements the Reranker interface using a Python subprocess
 * to run MLX models locally on Apple Silicon.
 */
export class MlxReranker implements Reranker {
    private config: MlxRerankerConfig;
    private ready: boolean = false;
    private pythonScriptPath: string;

    constructor(config: Partial<MlxRerankerConfig>) {
        // Apply defaults
        this.config = {
            provider: 'mlx', // This is a local provider
            model: config.model || 'jina-reranker-v3-mlx',
            modelPath: config.modelPath || '',
            pythonPath: config.pythonPath || 'python3',
            maxCandidates: config.maxCandidates || 50,
            topK: config.topK || 10,
            timeout: config.timeout || 60000, // Longer timeout for local inference
        };

        // Path to the Python script
        this.pythonScriptPath = join(__dirname, 'mlx_reranker.py');
    }

    /**
     * Initialize the MLX reranker
     * Checks if Python and MLX are available
     */
    async initialize(): Promise<void> {
        try {
            // Check if Python is available
            await this.runPythonScript(['--version']);
            
            // Check if MLX is installed
            await this.runPythonScript(['-c', 'import mlx; print(mlx.__version__)']);
            
            // Check if model path exists
            if (!this.config.modelPath) {
                throw new Error('MLX model path is required. Set modelPath in config.');
            }

            this.ready = true;
        } catch (error) {
            console.error('Failed to initialize MLX reranker:', error);
            this.ready = false;
            throw error;
        }
    }

    /**
     * Rerank documents using the MLX model
     * @param query - The search query
     * @param documents - Array of document contents to rerank
     * @param options - Optional reranking configuration
     * @returns Promise resolving to sorted reranking results
     */
    async rerank(
        query: string,
        documents: string[],
        options?: RerankOptions
    ): Promise<RerankResult[]> {
        if (!this.ready) {
            throw new Error('MLX reranker is not ready. Call initialize() first.');
        }

        if (documents.length === 0) {
            return [];
        }

        const topK = options?.topK ?? this.config.topK;
        const maxCandidates = Math.min(
            documents.length,
            options?.maxCandidates ?? this.config.maxCandidates
        );

        // Limit documents to maxCandidates
        const documentsToRerank = documents.slice(0, maxCandidates);

        try {
            const response = await this.runRerankingScript(
                query,
                documentsToRerank,
                topK
            );

            return response.results || [];
        } catch (error) {
            console.error('MLX reranking failed:', error);
            throw error;
        }
    }

    /**
     * Check if the reranker is ready to use
     * @returns True if the reranker is initialized and ready
     */
    isReady(): boolean {
        return this.ready;
    }

    /**
     * Get information about the reranker model
     * @returns Object containing model name and type
     */
    getModelInfo(): { name: string; type: 'api' | 'local' } {
        return {
            name: this.config.model,
            type: 'local',
        };
    }

    /**
     * Run the Python MLX reranker script
     * @param query - Search query
     * @param documents - Documents to rerank
     * @param topK - Number of top results to return
     * @returns Promise resolving to reranking results
     */
    private async runRerankingScript(
        query: string,
        documents: string[],
        topK: number
    ): Promise<MlxRerankerResponse> {
        return new Promise((resolve, reject) => {
            const args = [
                this.pythonScriptPath,
                '--query', query,
                '--model', this.config.modelPath,
                '--top-k', topK.toString(),
                '--documents', JSON.stringify(documents),
            ];

            const python = spawn(this.config.pythonPath!, args);

            let stdout = '';
            let stderr = '';

            python.stdout.on('data', (data) => {
                stdout += data.toString();
            });

            python.stderr.on('data', (data) => {
                stderr += data.toString();
            });

            python.on('close', (code) => {
                if (code !== 0) {
                    reject(new Error(`MLX reranker script exited with code ${code}: ${stderr}`));
                    return;
                }

                try {
                    const response: MlxRerankerResponse = JSON.parse(stdout);
                    
                    if (response.error) {
                        reject(new Error(response.error));
                        return;
                    }

                    resolve(response);
                } catch (parseError) {
                    reject(new Error(`Failed to parse MLX reranker response: ${parseError}\nOutput: ${stdout}`));
                }
            });

            python.on('error', (error) => {
                reject(new Error(`Failed to spawn Python process: ${error.message}`));
            });

            // Set timeout
            const timeout = setTimeout(() => {
                python.kill();
                reject(new Error(`MLX reranker timed out after ${this.config.timeout}ms`));
            }, this.config.timeout);

            python.on('close', () => {
                clearTimeout(timeout);
            });
        });
    }

    /**
     * Run a Python command and return the output
     * @param args - Arguments to pass to Python
     * @returns Promise resolving to the command output
     */
    private runPythonScript(args: string[]): Promise<string> {
        return new Promise((resolve, reject) => {
            const python = spawn(this.config.pythonPath!, args);

            let stdout = '';
            let stderr = '';

            python.stdout.on('data', (data) => {
                stdout += data.toString();
            });

            python.stderr.on('data', (data) => {
                stderr += data.toString();
            });

            python.on('close', (code) => {
                if (code !== 0) {
                    reject(new Error(`Python command failed with code ${code}: ${stderr}`));
                    return;
                }
                resolve(stdout);
            });

            python.on('error', (error) => {
                reject(new Error(`Failed to spawn Python process: ${error.message}`));
            });
        });
    }
}
